src/api/js/src/high-level.ts

```diff
@@ -0,0 +1,1362 @@
+import {
+  Z3_ast,
+  Z3_ast_kind,
+  Z3_ast_map,
+  Z3_ast_print_mode,
+  Z3_ast_vector,
+  Z3_context,
+  Z3_decl_kind,
+  Z3_error_code,
+  Z3_func_decl,
+  Z3_parameter_kind,
+  Z3_probe,
+  Z3_sort,
+  Z3_sort_kind,
+  Z3_symbol,
+  Z3_symbol_kind,
+  Z3_tactic
+} from "../build/wrapper";
+
+type ConfigParams = {
+  proof: boolean;
```

Comment:
This should either be auto-generated or obtained dynamically. The options can change and should only be updated in the main source.

Reply:
I added a function to the API to get descriptions of global parameters. 

https://github.com/Z3Prover/z3/blob/dd46224a1dd91c5bb7a3e12a52a23cf87441e6c1/src/api/z3_api.h#L1694

If you decide to bring ConfigParams back, then hopefully this feature can be used to populate contents dynamically.

Reply:
@NikolajBjorner is there any way of knowing more precisely which strings are accepted for a given parameter, or should they all just be typed as "string"?

Also, is there any way to get the default values for parameters? A way of getting the current value for the given context would be sufficient, since you can just make a context without setting any parameters and then read off the values for that context.

---

The fact that it has to be invoked at runtime rather than parsed from source means it will add another step to the build process, but we do have a thing we can invoke after calling `build:wasm`, so:

```js
'use strict';

(async () => {
  let mod = await require('./build/z3-built.js')();
  let ctx = mod._Z3_mk_context();
  let paramDescriptors = mod._Z3_get_global_param_descrs(ctx);
  mod._Z3_param_descrs_inc_ref(ctx, paramDescriptors);

  let configType = 'type ConfigParams = {';
  let size = mod._Z3_param_descrs_size(ctx, paramDescriptors);
  for (let i = 0; i < size; ++i) {
    let nameSymbol = mod._Z3_param_descrs_get_name(ctx, paramDescriptors, i);
    let name = mod.UTF8ToString(mod._Z3_get_symbol_string(ctx, nameSymbol));
    let comment = mod.UTF8ToString(mod._Z3_param_descrs_get_documentation(ctx, paramDescriptors, nameSymbol));
    configType += `\n  ${name}: string, // ${comment}`
  }
  configType += `\n};`;

  console.log(configType);
  mod.PThread.terminateAllThreads();
})().catch(e => {
  console.error(e);
  process.exit(1);
});
```
produces
```ts
type ConfigParams = {
  proof: string, // proof generation, it must be enabled when the Z3 context is created
  smtlib2_compliant: string, // enable/disable SMT-LIB 2.0 compliance
  timeout: string, // (default: infty) timeout in milliseconds.
  dump_models: string, // dump models whenever check-sat returns sat
  model_validate: string, // validate models produced by solvers
  model: string, // model generation for solvers, this parameter can be overwritten when creating a solver
  debug_ref_count: string, // debug support for AST reference counting
  encoding: string, // string encoding used internally: unicode|bmp|ascii
  trace: string, // trace generation for VCC
  trace_file_name: string, // trace out file name (see option 'trace')
  well_sorted_check: string, // type checker
  rlimit: string, // default resource limit used for solvers. Unrestricted when set to 0.
  unsat_core: string, // unsat-core generation for solvers, this parameter can be overwritten when creating a solver, not every solver in Z3 supports unsat core generation
  dot_proof_file: string, // file in which to output graphical proofs
  type_check: string, // type checker (alias for well_sorted_check)
  auto_config: string, // use heuristics to automatically select solver and configure it
  stats: string, // enable/disable statistics
  memory_max_alloc_count: string, // set hard upper limit for memory allocations, if 0 then there is no limit
  memory_high_watermark: string, // set high watermark for memory consumption (in megabytes), if 0 then there is no limit
  verbose: string, // be verbose, where the value is the verbosity level
  warning: string, // enable/disable warning messages
  memory_max_size: string, // set hard upper limit for memory consumption (in megabytes), if 0 then there is no limit
};
```

Reply:
> @NikolajBjorner is there any way of knowing more precisely which strings are accepted for a given parameter, or should they all just be typed as "string"

Z3 doesn't expose set of legal string options.

Reply:
> Also, is there any way to get the default values for parameters? A way of getting the current value for the given context would be sufficient, since you can just make a context without setting any parameters and then read off the values for that context.

You can get the parameter descriptions, then for each enumerated parameter retrieve the (current) parameter value.
The parameter defaults are not accessible as a separate value.

Reply:
How do you retrieve the current parameter value?

Reply:
> 
The fact that it has to be invoked at runtime rather than parsed from source means it will add another step to the build process, but we do have a thing we can invoke after calling build:wasm, so:

Right, this is not great. I thought that instead of exposing the parameters as a type "ConfigParams" as a map, it would be exposed as a closure that can be queried. Or if attributes can be created dynamically.

Reply:
I learned last week that we also have to update our documentation build to create an index of parameters and their descriptions.
The option z3.exe -pm (and similar) is available from the command-line for users, but I learned that a (power) user spent quite some time first browsing the documentation (that needs other facelift) for a description of parameters.
My own behavior is far from representative: I just browse the source code.